data/createPages/createPageVariants.js

```diff
@@ -22,13 +22,17 @@ const addLanguagesToSet =
     }
   };
 
+const getLanguagesFromContent = (contentOrderedList) => {
+  const languageSet = new Set();
+  contentOrderedList.forEach(addLanguagesToSet(languageSet));
```

Comment:
Sorry, I forgot to respond here:

Short answer is we get these languages from the HTML attributes on the page. This provides a substitute set of languages to choose from for users in case the languages specified in the frontmatter meta are not set.

Long answer, in the interest of providing more insight into the full pipeline:

addLanguagesToSet is a recursive function that travels down the content data tree that looks very much like this:

`type ContentData = { data: ContentData[] | string | null; attribs?: { ...htmlAttributes }; type: HtmlElementName; }`

so the HTML would look like:

`<p lang="javascript">Some content</p><p lang="csharp">Some content</p><p lang="javascript">Some different content</p>`

and the data would look like:

`[{ data: "Some content", type: "p", attribs: { lang: "javascript" } }, { data: "Some content", type: "p", attribs: { lang: "csharp" } }, { data: "Some different content", type: "p", attribs: { lang: "javascript" } }]`

so in this case the unduplicated array would look like:

`['javascript', 'csharp', 'javascript']`

Reply:
thanks for the explanation @MymmiJ :+1: